Localized new strings added in the redesigned Material Date Picker

darrenaustin opened the pull request:
## Description

The new redesign of the Material Date Picker (#50546) and the addition of the Date Range Picker (#55939) added several new user facing strings that need to be localized. This PR is the first step for that localization. It adds the new strings to the `MaterialLocalizations` class with english defaults in the `DefaultMaterialLocalizations` class. The locale .arb files have all been update with the english values for these strings. When the translations are available another PR will be created to complete the process.

## Tests

I added tests for all the new strings to `localizations_test.dart` (including a few that weren't part of this work, but were missing from the file 😉 ).

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them?

- [x] No, no existing tests failed, so this is *not* a breaking change.

darrenaustin committed:
Added new user facing strings needed for the Date Pickers to the material localization system.

rami-a reviewed (approved):
LGTM

shihaohong commented on packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:
```diff
@@ -543,10 +543,9 @@ class _MonthItem extends StatelessWidget {
     if (isSelectedDayStart) {
-      // TODO(darrenaustin): localize 'Start Date' and 'End Date'
-      semanticLabel = 'Start Date ' + semanticLabel;
+      semanticLabel = '${localizations.dateRangeStartLabel} $semanticLabel';
```
Do we know that for non-English messages, this order of the strings would be desirable? (ie. `Start Date` comes before `$semanticLabel`). It might be important to share some context on words/strings regarding the label (what might come before/after the string), as well as whether it will be read aloud, visually read, or both (this came up as a question recently with the Cupertino tabs semantics label).

darrenaustin replied:
Good point. I just changed these two labels to parameterized messages, so they can adjust the order as they like.

shihaohong commented on packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:
```diff
@@ -543,10 +543,9 @@ class _MonthItem extends StatelessWidget {
+      semanticLabel = '${localizations.dateRangeStartLabel} $semanticLabel';
     } else if (isSelectedDayEnd) {
-      semanticLabel = 'End Date ' + semanticLabel;
+      semanticLabel = '${localizations.dateRangeEndLabel} $semanticLabel';
```
Ditto w/ above question

shihaohong commented on packages/flutter/lib/src/material/pickers/calendar_date_picker.dart:
```diff
@@ -366,8 +366,7 @@ class _DatePickerModeToggleButtonState extends State<_DatePickerModeToggleButton
             child: Semantics(
-              // TODO(darrenaustin): localize 'Select year'
-              label: 'Select year',
+              label: MaterialLocalizations.of(context).selectYearLabel,
```
Might be worth naming/mentioning that this is a semantics label that will be used for VoiceOver and Talkback.
```suggestion
              label: MaterialLocalizations.of(context).selectYearSemanticsLabel,
```

darrenaustin replied:
Renamed.

shihaohong commented on packages/flutter_localizations/lib/src/l10n/material_en.arb:
```diff
@@ -173,6 +173,86 @@
     "description": "Label read out by accessibility tools (TalkBack or VoiceOver) for a modal barrier to indicate that a tap dismisses the barrier. A modal barrier can for example be found behind a alert or popup to block user interaction with elements behind it."
   },
 
+  "dateSeparator": "/",
```
For the `_en.arb` descriptions, could we specify if the label is going to be a visual element, a semantics label, or both? I recently had translators ask me about this with the Cupertino tabs semantics label and the translations for certain languages will be unique based on that context.

darrenaustin replied:
I was just following the pattern I saw in the existing labels in the file. They are all visual elements unless described as an 'accessibility label' or 'accessibility announcement'.

shihaohong reviewed (commented):
The code logic looks good to me, just a few questions about semantics vs visual labels based on some recent similar experiences I've had

darrenaustin committed:
Updates from code review
- Changed the TBD entries to the english strings.
- Made a couple of the semantic labels more customizable in the localization by parameterizing them.

shihaohong reviewed (approved):
LGTM! Definitely much better that the untranslated values are now the English text instead of 'TBD' since it won't be a breaking change.

shihaohong commented:
Also, small detail, but we might want to update the description section of the PR's original post since the strings are no longer placeholder 'TBD's anymore
